### DISCUSSION TITLE
Markdown extensions on a per page basis

### DISCUSSION THREAD

**Discussion by rodrigoalcarazdelaosa** [mkdocs/mkdocs#3715] [Q&A]
Basically what the title says. We can specify which markdown extensions to use in our `mkdocs.yml` file and they'll be applied to our whole docs. Is there any way to specify which markdown extensions we want to use on a per page basis (putting them in the front matter)? Thanks!

**Comment by kchr** [comment]
Would you mind sharing an example case for when this would be useful? There might be other solutions to your problem, and more context is always helpful when it comes to feature requests.

**Reply by rodrigoalcarazdelaosa** [reply]
This originated [here](https://github.com/squidfunk/mkdocs-material/discussions/3813#discussioncomment-9546575). [Material for MkDocs](https://squidfunk.github.io/mkdocs-material/) allows us for example to enable the [Table of Contents](https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown/#table-of-contents) extension via our `mkdocs.yml` file. That way we can add anchor links to our headings. But what if we want to disable that extension on a specific page. Having the possibility to enable extensions on a per page would be welcome. I can imagine another example with [math extensions](https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown-extensions/#arithmatex), which we could enable only on pages that actually use math, saving some JS performance.

**Comment by kamilkrzyskow** [comment]
I guess, this question about changing global markdown extensions is a bit similar to this question about changing global features:
- https://github.com/squidfunk/mkdocs-material/discussions/6865

where I suggested on using a hook and modifying the global object between pages. It requires storing the original and replacing the global value between events. This approach can be definitely extended to Markdown extensions, but note that `mdx_configs` are separate from `markdown_extensiosn` internally:

https://github.com/mkdocs/mkdocs/blob/fb1d106747930b094bc464189d4b1c349fc541c6/mkdocs/config/defaults.py#L132-L138

The Markdown processor is recreated between pages, so only the initializing values need to be changed between pages.
Also as here we want to modify the markdown extensions, `on_page_markdown` and `on_page_content` events should be used, instead of the ones in the other discussion above.
A possible reference on changing the settings: https://github.com/mkdocs/mkdocs/discussions/3196

> Note that there are talks about changing the logic for rendering pages, so perhaps such hacky ways might stop working in the future.
> - https://github.com/mkdocs/mkdocs/issues/3695
